dhcp/dhcp.go

```diff
@@ -0,0 +1,446 @@
+	if innerErr := unix.SetsockoptTimeval(fd, unix.SOL_SOCKET, unix.SO_RCVTIMEO, &timeval); innerErr != nil {
+		// don't leak sockets
+		if closeErr := ret.Close(); closeErr != nil {
+			return nil, errors.Wrap(innerErr, "could not set timeout on socket and failed to close read socket")
+		}
+		return nil, errors.Wrap(innerErr, "could not set timeout on socket")
+	}
+
+	return ret, nil
+}
+
+func (s *Socket) Read(p []byte) (n int, err error) {
+	n, _, innerErr := unix.Recvfrom(s.fd, p, 0)
+	if innerErr != nil {
+		return 0, errors.Wrap(err, "failed unix recv from")
+	}
+	return n, nil
+}
+
+func (s *Socket) Close() error {
+	// Ensure the file descriptor is closed when done
+	if err := unix.Close(s.fd); err != nil {
+		return errors.Wrap(err, "error closing dhcp unix socket")
+	}
+	return nil
+}
+
+// GenerateTransactionID generates a random 32-bits number suitable for use as TransactionID
+func GenerateTransactionID() (TransactionID, error) {
+	var xid TransactionID
+	_, err := rand.Read(xid[:])
+	if err != nil {
+		return xid, errors.Errorf("could not get random number: %v", err)
+	}
+	return xid, nil
+}
+
+func makeListeningSocket(ifname string) (int, error) {
+	fd, err := unix.Socket(unix.AF_PACKET, unix.SOCK_DGRAM, int(htons(unix.ETH_P_IP)))
+	if err != nil {
+		return fd, errors.Wrap(err, "dhcp socket creation failure")
+	}
+	iface, err := net.InterfaceByName(ifname)
+	if err != nil {
+		return fd, errors.Wrap(err, "dhcp failed to get interface")
+	}
+	llAddr := unix.SockaddrLinklayer{
+		Ifindex:  iface.Index,
+		Protocol: htons(unix.ETH_P_IP),
+	}
+	err = unix.Bind(fd, &llAddr)
+	return fd, errors.Wrap(err, "dhcp failed to bind")
+}
+
+// MakeBroadcastSocket creates a socket that can be passed to unix.Sendto
+// that will send packets out to the broadcast address.
+func MakeBroadcastSocket(ifname string) (int, error) {
+	fd, err := makeRawSocket(ifname)
+	if err != nil {
+		return fd, err
+	}
+	err = unix.SetsockoptInt(fd, unix.SOL_SOCKET, unix.SO_BROADCAST, 1)
+	if err != nil {
+		return fd, errors.Wrap(err, "dhcp failed to set sockopt")
+	}
+	return fd, nil
+}
+
+func htons(v uint16) uint16 {
+	var tmp [2]byte
+	binary.BigEndian.PutUint16(tmp[:], v)
+	return binary.LittleEndian.Uint16(tmp[:])
+}
+
+func BindToInterface(fd int, ifname string) error {
+	return errors.Wrap(unix.BindToDevice(fd, ifname), "failed to bind to device")
+}
+
+// makeRawSocket creates a socket that can be passed to unix.Sendto.
+func makeRawSocket(ifname string) (int, error) {
+	fd, err := unix.Socket(unix.AF_INET, unix.SOCK_RAW, unix.IPPROTO_RAW)
+	if err != nil {
+		return fd, errors.Wrap(err, "dhcp raw socket creation failure")
+	}
+	err = unix.SetsockoptInt(fd, unix.SOL_SOCKET, unix.SO_REUSEADDR, 1)
+	if err != nil {
+		return fd, errors.Wrap(err, "dhcp failed to set raw sockopt")
+	}
+	err = unix.SetsockoptInt(fd, unix.IPPROTO_IP, unix.IP_HDRINCL, 1)
+	if err != nil {
+		return fd, errors.Wrap(err, "dhcp failed to set second raw sockopt")
+	}
```

Comment:
why it requires to set these options?

Reply:
REUSEADDR is to avoid "address already in use" errors.

IP_HDR_INCL is because I believe later on when we write to this broadcast socket, our packet _already_ contains the header (we create it with MakeRawUDPPacket). Omitting this I believe would mean another header would be added to the packet, which we don't want.

Reply:
i dont think `REUSEADDR` is required. can we test without it?